nixos/modules/services/monitoring/dcgm.nix

```diff
@@ -0,0 +1,208 @@
+        Usually this should be set to the default nvidia-dcgm account.
+      '';
+    };
+
+    homeDir = mkOption {
+      type = types.path;
+      default = "/var/lib/dcgm";
+      description = lib.mdDoc ''
+        The full path to the directory that the DCGM diagnostic should be launched from.
+        This becomes the default output directory for DCGM diagnostic log and stats files.
+      '';
+    };
+
+    openFirewall = mkOption {
+      type = types.bool;
+      default = false;
+      description = lib.mdDoc ''
+        Open the firewall for the port.
+      '';
+    };
+  };
+
+  config = mkMerge [
+    (mkIf cfg.enable {
+      # TODO(@connorbaker): Look at
+      # https://github.com/nix-community/queued-build-hook/blob/6a69b97d2951e3440f209930cde0464d7754274c/module.nix
+      # for an example service which uses a socket.
+
+      # TODO(@connorbaker): Currently complaining about being run as non-root. We can't run it as the service account?
+
+      # TODO(@connorbaker): Current errors: (get new copy)
+
+      systemd.services.dcgm = {
+        description = "NVIDIA Data Center GPU Manager (DCGM)";
+        wantedBy = ["multi-user.target"];
+        after = ["network.target"];
+        environment = {
+          # NOTE: Add the libnvidia-ml.so.1 and libnvidia_nscq.so libraries to the LD_LIBRARY_PATH
+          # using the current NVIDIA driver.
+          # If we don't we get a failure here:
+          # https://github.com/NVIDIA/DCGM/blob/7e1012302679e4bb7496483b32dcffb56e528c92/sdk/nvidia/nvml/nvml_loader/nvml_loader.cpp#L199-L212.
+          # The underlying problem is that DCGM expects the library to be in certain paths:
+          # https://github.com/NVIDIA/DCGM/blob/7e1012302679e4bb7496483b32dcffb56e528c92/sdk/nvidia/nvml/nvml_loader/nvml_loader.cpp#L126-L131.
+          LD_LIBRARY_PATH = "${config.hardware.nvidia.package}/lib";
+          DCGM_HOME_DIR = cfg.homeDir;
+        };
+        serviceConfig = {
+          # TODO(@connorbaker): Upstream's systemd config template is here: https://github.com/NVIDIA/DCGM/blob/7e1012302679e4bb7496483b32dcffb56e528c92/config-files/systemd/nvidia-dcgm.service.in
+          # TODO(@connorbaker): Are these relevant?
+          # AmbientCapabilities = ["CAP_SYS_ADMIN"];
+          # CapabilityBoundingSet = ["CAP_SYS_ADMIN"];
+
+          WorkingDirectory = cfg.homeDir;
+          # User = cfg.serviceAccount;
+          # Group = cfg.serviceAccount;
+          # TODO(@connorbaker): What does systemd say the different between runtime and working directory are?
+          # RuntimeDirectory = cfg.serviceAccount;
+          # RuntimeDirectoryMode = "0755";
+
+          # Restart = "always";
+          # RestartSec = "5s";
+          # TODO(@connorbaker): What is DynamicUser?
+          # DynamicUser = true;
+          ExecStart = lib.concatStringsSep " \\\n  " (
+            [
+              "${lib.getBin pkgs.dcgm}/bin/nv-hostengine"
```

Comment:
Perhaps the `cfg.package` (which defaults to `pkgs.dcgm`) would be better? No strong opinion from me.